From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Rendy Sebpian <nightcoremosta@gmail.com>
Date: Sun, 26 Jul 2026 23:15:02 +0700
Subject: [PATCH 1/4] Integrate the SceneFX renderer


diff --git a/meson.build b/meson.build
index d85823c..c323b5b 100644
--- a/meson.build
+++ b/meson.build
@@ -56,6 +56,12 @@ wlroots = dependency(
   version: ['>=0.20.1', '<0.21.0'],
 )
 
+scenefx = dependency('scenefx-0.5')
+add_project_arguments(
+  '-DSCENEFX_VERSION="@0@"'.format(scenefx.version()),
+  language: 'c',
+)
+
 wlroots_has_xwayland = wlroots.get_variable('have_xwayland') == 'true'
 have_libsfdo = not get_option('icon').disabled()
 
@@ -161,6 +167,7 @@ subdir('protocols')
 labwc_deps = [
   server_protos,
   wayland_server,
+  scenefx,
   wlroots,
   xkbcommon,
 	xcb_ewmh,
diff --git a/src/action.c b/src/action.c
index 265ff6e..412d25a 100644
--- a/src/action.c
+++ b/src/action.c
@@ -7,7 +7,7 @@
 #include <strings.h>
 #include <unistd.h>
 #include <wlr/types/wlr_cursor.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "action-prompt-codes.h"
 #include "common/buf.h"
diff --git a/src/common/lab-scene-rect.c b/src/common/lab-scene-rect.c
index 6689d48..99309c5 100644
--- a/src/common/lab-scene-rect.c
+++ b/src/common/lab-scene-rect.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include "common/lab-scene-rect.h"
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/mem.h"
 #include "common/scene-helpers.h"
 
diff --git a/src/common/scene-helpers.c b/src/common/scene-helpers.c
index e9c80bb..4051346 100644
--- a/src/common/scene-helpers.c
+++ b/src/common/scene-helpers.c
@@ -3,7 +3,7 @@
 #include "common/scene-helpers.h"
 #include <assert.h>
 #include <wlr/types/wlr_output.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "common/mem.h"
 #include "magnifier.h"
diff --git a/src/cycle/cycle.c b/src/cycle/cycle.c
index 0f183dd..e68c027 100644
--- a/src/cycle/cycle.c
+++ b/src/cycle/cycle.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include "cycle.h"
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/box.h>
 #include <wlr/util/log.h>
 #include "common/lab-scene-rect.h"
diff --git a/src/cycle/osd-classic.c b/src/cycle/osd-classic.c
index 7e5abfb..302fcee 100644
--- a/src/cycle/osd-classic.c
+++ b/src/cycle/osd-classic.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <assert.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/box.h>
 #include <wlr/util/log.h>
 #include "common/buf.h"
diff --git a/src/cycle/osd-scroll.c b/src/cycle/osd-scroll.c
index e788a47..e2cfe06 100644
--- a/src/cycle/osd-scroll.c
+++ b/src/cycle/osd-scroll.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/lab-scene-rect.h"
 #include "common/scene-helpers.h"
 #include "labwc.h"
diff --git a/src/cycle/osd-thumbnail.c b/src/cycle/osd-thumbnail.c
index defedf0..ad99517 100644
--- a/src/cycle/osd-thumbnail.c
+++ b/src/cycle/osd-thumbnail.c
@@ -4,7 +4,7 @@
 #include <wlr/render/swapchain.h>
 #include <wlr/types/wlr_buffer.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "config/rcxml.h"
 #include "common/box.h"
 #include "common/buf.h"
@@ -73,6 +73,11 @@ render_node(struct wlr_render_pass *pass,
 		/* should be unreached */
 		wlr_log(WLR_ERROR, "ignoring rect");
 		break;
+	case WLR_SCENE_NODE_SHADOW:
+	case WLR_SCENE_NODE_OPTIMIZED_BLUR:
+	case WLR_SCENE_NODE_BLUR:
+		/* Effects are not part of the window-content thumbnail. */
+		break;
 	}
 }
 
diff --git a/src/debug.c b/src/debug.c
index fef4d81..76b8370 100644
--- a/src/debug.c
+++ b/src/debug.c
@@ -2,7 +2,7 @@
 #include "debug.h"
 #include <stdlib.h>
 #include <wlr/types/wlr_layer_shell_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/lab-scene-rect.h"
 #include "common/scene-helpers.h"
 #include "common/string-helpers.h"
@@ -42,6 +42,12 @@ get_node_type(struct wlr_scene_node *node)
 			return "surface";
 		}
 		return "buffer";
+	case WLR_SCENE_NODE_SHADOW:
+		return "shadow";
+	case WLR_SCENE_NODE_OPTIMIZED_BLUR:
+		return "optimized-blur";
+	case WLR_SCENE_NODE_BLUR:
+		return "blur";
 	}
 	return "error";
 }
diff --git a/src/desktop.c b/src/desktop.c
index 57ef9e3..a73fad1 100644
--- a/src/desktop.c
+++ b/src/desktop.c
@@ -4,7 +4,7 @@
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_layer_shell_v1.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/types/wlr_subcompositor.h>
 #include <wlr/types/wlr_xdg_shell.h>
diff --git a/src/dnd.c b/src/dnd.c
index 5d7d050..a0c21d7 100644
--- a/src/dnd.c
+++ b/src/dnd.c
@@ -2,7 +2,7 @@
 #include "dnd.h"
 #include <assert.h>
 #include <wlr/types/wlr_data_device.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "common/scene-helpers.h"
 #include "config/rcxml.h"
diff --git a/src/edges.c b/src/edges.c
index 4e5d7ca..fefa785 100644
--- a/src/edges.c
+++ b/src/edges.c
@@ -4,7 +4,7 @@
 #include <limits.h>
 #include <pixman.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/box.h>
 #include "common/border.h"
 #include "common/box.h"
diff --git a/src/input/cursor.c b/src/input/cursor.c
index e8ac7d9..fb69281 100644
--- a/src/input/cursor.c
+++ b/src/input/cursor.c
@@ -11,7 +11,7 @@
 #include <wlr/types/wlr_pointer_constraints_v1.h>
 #include <wlr/types/wlr_primary_selection.h>
 #include <wlr/types/wlr_relative_pointer_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_subcompositor.h>
 #include <wlr/types/wlr_xcursor_manager.h>
 #include <wlr/types/wlr_xdg_shell.h>
diff --git a/src/input/ime.c b/src/input/ime.c
index 1758d1c..8b6646c 100644
--- a/src/input/ime.c
+++ b/src/input/ime.c
@@ -6,7 +6,7 @@
 #include <wlr/types/wlr_input_method_v2.h>
 #include <wlr/types/wlr_layer_shell_v1.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/types/wlr_text_input_v3.h>
 #include <wlr/types/wlr_virtual_keyboard_v1.h>
diff --git a/src/input/key-state.c b/src/input/key-state.c
index 6d08989..bbe002a 100644
--- a/src/input/key-state.c
+++ b/src/input/key-state.c
@@ -5,7 +5,7 @@
 #include <stdio.h>
 #include <stdlib.h>
 #include <wlr/types/wlr_keyboard.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <xkbcommon/xkbcommon.h>
 #include "config/rcxml.h"
 #include "common/buf.h"
diff --git a/src/input/tablet.c b/src/input/tablet.c
index 82d5e80..68a3258 100644
--- a/src/input/tablet.c
+++ b/src/input/tablet.c
@@ -7,7 +7,7 @@
 #include <wlr/types/wlr_tablet_tool.h>
 #include <wlr/types/wlr_tablet_v2.h>
 #include <wlr/util/log.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "common/macros.h"
 #include "common/mem.h"
diff --git a/src/input/touch.c b/src/input/touch.c
index 42fab35..6491d30 100644
--- a/src/input/touch.c
+++ b/src/input/touch.c
@@ -2,7 +2,7 @@
 #include "input/touch.h"
 #include <wayland-util.h>
 #include <wlr/types/wlr_cursor.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/types/wlr_touch.h>
 #include <linux/input-event-codes.h>
diff --git a/src/layers.c b/src/layers.c
index eaa93f2..4a2e757 100644
--- a/src/layers.c
+++ b/src/layers.c
@@ -14,7 +14,7 @@
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_fractional_scale_v1.h>
 #include <wlr/types/wlr_layer_shell_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_xdg_shell.h>
 #include <wlr/util/log.h>
 #include "common/macros.h"
diff --git a/src/magnifier.c b/src/magnifier.c
index 43a7174..3652a95 100644
--- a/src/magnifier.c
+++ b/src/magnifier.c
@@ -6,7 +6,7 @@
 #include <wlr/render/swapchain.h>
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_output.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/transform.h>
 #include "common/box.h"
 #include "config/rcxml.h"
diff --git a/src/main.c b/src/main.c
index 081bd62..e0a4a32 100644
--- a/src/main.c
+++ b/src/main.c
@@ -70,7 +70,7 @@ static void
 print_version(void)
 {
 	#define FEATURE_ENABLED(feature) (HAVE_##feature ? "+" : "-")
-	printf("labwc %s (%sxwayland %snls %srsvg %slibsfdo) wlroots-%d.%d.%d\n",
+	printf("labwc %s (%sxwayland %snls %srsvg %slibsfdo) wlroots-%d.%d.%d scenefx-%s\n",
 		LABWC_VERSION,
 		FEATURE_ENABLED(XWAYLAND),
 		FEATURE_ENABLED(NLS),
@@ -78,7 +78,8 @@ print_version(void)
 		FEATURE_ENABLED(LIBSFDO),
 		wlr_version_get_major(),
 		wlr_version_get_minor(),
-		wlr_version_get_micro()
+		wlr_version_get_micro(),
+		SCENEFX_VERSION
 	);
 	#undef FEATURE_ENABLED
 }
diff --git a/src/menu/menu.c b/src/menu/menu.c
index c0e93c5..ced7fb3 100644
--- a/src/menu/menu.c
+++ b/src/menu/menu.c
@@ -10,7 +10,7 @@
 #include <string.h>
 #include <strings.h>
 #include <unistd.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_xdg_shell.h>
 #include <wlr/util/log.h>
 #include "action.h"
diff --git a/src/node.c b/src/node.c
index 19b4b5a..c7ffd40 100644
--- a/src/node.c
+++ b/src/node.c
@@ -2,7 +2,7 @@
 #include "node.h"
 #include <assert.h>
 #include <stdlib.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/mem.h"
 #include "ssd.h"
 
diff --git a/src/output.c b/src/output.c
index 2eab8ec..fd27f2c 100644
--- a/src/output.c
+++ b/src/output.c
@@ -22,7 +22,7 @@
 #include <wlr/types/wlr_output_power_management_v1.h>
 #include <wlr/types/wlr_xcursor_manager.h>
 #include <wlr/types/wlr_xdg_output_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "common/macros.h"
 #include "common/mem.h"
diff --git a/src/overlay.c b/src/overlay.c
index 817bb3f..63d28f8 100644
--- a/src/overlay.c
+++ b/src/overlay.c
@@ -3,7 +3,7 @@
 #include <assert.h>
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/lab-scene-rect.h"
 #include "config/rcxml.h"
 #include "labwc.h"
diff --git a/src/resize-outlines.c b/src/resize-outlines.c
index 113f128..eba7f68 100644
--- a/src/resize-outlines.c
+++ b/src/resize-outlines.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include "resize-outlines.h"
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/border.h"
 #include "common/lab-scene-rect.h"
 #include "config/rcxml.h"
diff --git a/src/scaled-buffer/scaled-buffer.c b/src/scaled-buffer/scaled-buffer.c
index 95ec896..119b212 100644
--- a/src/scaled-buffer/scaled-buffer.c
+++ b/src/scaled-buffer/scaled-buffer.c
@@ -6,7 +6,7 @@
 #include <wayland-server-core.h>
 #include <wlr/types/wlr_buffer.h>
 #include <wlr/types/wlr_output.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/log.h>
 #include "buffer.h"
 #include "common/list.h"
diff --git a/src/server.c b/src/server.c
index d0c8c04..940e585 100644
--- a/src/server.c
+++ b/src/server.c
@@ -2,6 +2,7 @@
 #define _POSIX_C_SOURCE 200809L
 #include "config.h"
 #include <assert.h>
+#include <scenefx/render/fx_renderer/fx_renderer.h>
 #include <signal.h>
 #include <string.h>
 #include <sys/wait.h>
@@ -30,7 +31,7 @@
 #include <wlr/types/wlr_presentation_time.h>
 #include <wlr/types/wlr_primary_selection_v1.h>
 #include <wlr/types/wlr_relative_pointer_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_screencopy_v1.h>
 #include <wlr/types/wlr_security_context_v1.h>
 #include <wlr/types/wlr_single_pixel_buffer_v1.h>
@@ -373,7 +374,7 @@ handle_renderer_lost(struct wl_listener *listener, void *data)
 {
 	wlr_log(WLR_INFO, "Re-creating renderer after GPU reset");
 
-	struct wlr_renderer *renderer = wlr_renderer_autocreate(server.backend);
+	struct wlr_renderer *renderer = fx_renderer_create(server.backend);
 	if (!renderer) {
 		wlr_log(WLR_ERROR, "Unable to create renderer");
 		return;
@@ -527,7 +528,7 @@ server_init(void)
 	 * The renderer is responsible for defining the various pixel formats it
 	 * supports for shared memory, this configures that for clients.
 	 */
-	server.renderer = wlr_renderer_autocreate(server.backend);
+	server.renderer = fx_renderer_create(server.backend);
 	if (!server.renderer) {
 		wlr_log(WLR_ERROR, "unable to create renderer");
 		exit(EXIT_FAILURE);
diff --git a/src/session-lock.c b/src/session-lock.c
index 4e5eb8a..21d3e3d 100644
--- a/src/session-lock.c
+++ b/src/session-lock.c
@@ -3,7 +3,7 @@
 #include "session-lock.h"
 #include <assert.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_session_lock_v1.h>
 #include "common/mem.h"
 #include "common/scene-helpers.h"
diff --git a/src/ssd/resize-indicator.c b/src/ssd/resize-indicator.c
index fa1166a..b904d9a 100644
--- a/src/ssd/resize-indicator.c
+++ b/src/ssd/resize-indicator.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/util/box.h>
 #include <wlr/util/log.h>
 #include "common/scene-helpers.h"
diff --git a/src/ssd/ssd-border.c b/src/ssd/ssd-border.c
index c2c99f8..13d47f9 100644
--- a/src/ssd/ssd-border.c
+++ b/src/ssd/ssd-border.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/macros.h"
 #include "common/scene-helpers.h"
 #include "config/rcxml.h"
diff --git a/src/ssd/ssd-button.c b/src/ssd/ssd-button.c
index 7342dfa..9f9c7ea 100644
--- a/src/ssd/ssd-button.c
+++ b/src/ssd/ssd-button.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "config/rcxml.h"
 #include "common/list.h"
 #include "common/mem.h"
diff --git a/src/ssd/ssd-extents.c b/src/ssd/ssd-extents.c
index 3840b10..797c803 100644
--- a/src/ssd/ssd-extents.c
+++ b/src/ssd/ssd-extents.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include <pixman.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/scene-helpers.h"
 #include "config/rcxml.h"
 #include "labwc.h"
diff --git a/src/ssd/ssd-shadow.c b/src/ssd/ssd-shadow.c
index 9a45612..bda6204 100644
--- a/src/ssd/ssd-shadow.c
+++ b/src/ssd/ssd-shadow.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 
 #include <assert.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "buffer.h"
 #include "common/scene-helpers.h"
 #include "config/rcxml.h"
diff --git a/src/ssd/ssd-titlebar.c b/src/ssd/ssd-titlebar.c
index d1a0881..aa9acc9 100644
--- a/src/ssd/ssd-titlebar.c
+++ b/src/ssd/ssd-titlebar.c
@@ -4,7 +4,7 @@
 #include <assert.h>
 #include <string.h>
 #include <wlr/render/pixman.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "buffer.h"
 #include "common/mem.h"
 #include "common/scene-helpers.h"
diff --git a/src/ssd/ssd.c b/src/ssd/ssd.c
index e2ab637..4a0d5e6 100644
--- a/src/ssd/ssd.c
+++ b/src/ssd/ssd.c
@@ -10,7 +10,7 @@
 #include <assert.h>
 #include <strings.h>
 #include <wlr/types/wlr_cursor.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "common/mem.h"
 #include "common/scene-helpers.h"
 #include "config/rcxml.h"
diff --git a/src/view.c b/src/view.c
index 21005f2..c0c4cc5 100644
--- a/src/view.c
+++ b/src/view.c
@@ -1,11 +1,11 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include "view.h"
 #include <assert.h>
+#include <scenefx/types/wlr_scene.h>
 #include <strings.h>
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_keyboard_group.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
 #include <wlr/types/wlr_security_context_v1.h>
 #include <wlr/types/wlr_xdg_shell.h>
 #include "action.h"
diff --git a/src/workspaces.c b/src/workspaces.c
index a1ed911..47e8c94 100644
--- a/src/workspaces.c
+++ b/src/workspaces.c
@@ -10,7 +10,7 @@
 #include <strings.h>
 #include <wlr/types/wlr_ext_workspace_v1.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include "buffer.h"
 #include "common/font.h"
 #include "common/graphic-helpers.h"
diff --git a/src/xdg-popup.c b/src/xdg-popup.c
index 7ca9ab0..feee37b 100644
--- a/src/xdg-popup.c
+++ b/src/xdg-popup.c
@@ -7,7 +7,7 @@
  *	- keeping non-layer-shell xdg-popups outside the layers.c code
  */
 
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_xdg_shell.h>
 #include "common/macros.h"
 #include "common/mem.h"
diff --git a/src/xdg.c b/src/xdg.c
index ccba0de..d81c2e0 100644
--- a/src/xdg.c
+++ b/src/xdg.c
@@ -3,7 +3,7 @@
 #include <assert.h>
 #include <wlr/types/wlr_cursor.h>
 #include <wlr/types/wlr_fractional_scale_v1.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_xdg_activation_v1.h>
 #include <wlr/types/wlr_xdg_dialog_v1.h>
 #include <wlr/types/wlr_xdg_shell.h>
diff --git a/src/xwayland-unmanaged.c b/src/xwayland-unmanaged.c
index 89d40e8..917f6db 100644
--- a/src/xwayland-unmanaged.c
+++ b/src/xwayland-unmanaged.c
@@ -1,7 +1,7 @@
 // SPDX-License-Identifier: GPL-2.0-only
 #include <assert.h>
 #include <wlr/types/wlr_compositor.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/xwayland.h>
 #include "common/list.h"
diff --git a/src/xwayland.c b/src/xwayland.c
index 5838d41..a9f2ef6 100644
--- a/src/xwayland.c
+++ b/src/xwayland.c
@@ -5,7 +5,7 @@
 #include <stdlib.h>
 #include <wlr/types/wlr_compositor.h>
 #include <wlr/types/wlr_output_layout.h>
-#include <wlr/types/wlr_scene.h>
+#include <scenefx/types/wlr_scene.h>
 #include <wlr/types/wlr_seat.h>
 #include <wlr/types/wlr_xcursor_manager.h>
 #include <wlr/xwayland.h>
